components/places/src/storage/history.rs

```diff
@@ -446,7 +546,31 @@ pub mod history_sync {
             }
             None => new_page_info(db, &url, Some(incoming_guid.clone()))?,
         };
+
+        // Skip visits that are in tombstones, or that happen at the same time
+        // as visit that's already present. The 2nd lets us avoid inserting
+        // visits that we sent up to the server in the first place.
+        //
+        // It does cause us to ignore visits that legitimately happen
+        // at the same time, but that's probably fine and not worth
+        // worrying about.
+        let visits_to_skip: HashSet<Timestamp> = db.query_rows_into_cached(
```

Comment:
I think we could lift this query into the `INSERT` in `add_visit` (`INSERT...SELECT...WHERE NOT EXISTS`), but it's probably not worth finagling, given that (IIUC) we don't expect this to happen too often...

Reply:
It might be worth doing if we inserted them all in one go, but I suspect moving it into the query would cause sqlite to repeatedly have to do this part of it for each visit inserted.